src/Week5/HW05.hs

```diff
@@ -0,0 +1,144 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE OverloadedStrings, RecordWildCards #-}
+module Week5.HW05 where
+
+import Data.List
+import Data.ByteString.Lazy (ByteString)
+import Data.Map.Strict (Map)
+import System.Environment (getArgs)
+
+import qualified Data.ByteString.Lazy as BS
+import qualified Data.Map.Strict as Map
+
+import Data.Bits ( Bits(xor) )
+
+import Week5.Parser
+
+-- Exercise 1 -----------------------------------------
+
+getSecret :: FilePath -> FilePath -> IO ByteString
+getSecret path1 path2 =
+  do
+    f1 <- BS.readFile path1
+    f2 <- BS.readFile path2
+    return $ BS.filter (/=0) $ BS.pack $ BS.zipWith xor f1 f2
+
+
+
+-- Exercise 2 -----------------------------------------
+
+decryptWithKey :: ByteString -> FilePath -> IO ()
+decryptWithKey key output =
+  do
+    input <- BS.readFile $ output ++ ".enc"
+    BS.writeFile output $ BS.pack $ BS.zipWith xor input $ BS.cycle key
+
+
+-- Exercise 3 -----------------------------------------
+
+parseFile :: FromJSON a => FilePath -> IO (Maybe a)
+parseFile fp = do
+  f <- BS.readFile fp
+  return $ decode f
+
+-- Exercise 4 -----------------------------------------
+
+getBadTs :: FilePath -> FilePath -> IO (Maybe [Transaction])
+getBadTs victimPath transactionPath =
+  do
+    vf <- parseFile victimPath
+    tf <- parseFile transactionPath
+    let res = case (vf, tf) of 
+          (Nothing , _) -> Nothing 
+          (_, Nothing ) -> Nothing 
+          (Just v, Just t) -> Just $ helper v t 
+    return res
```

Comment:
Not sure if you saw applicative functors already, but this could be simplified using those like this (because maybe is applicative functor too):

```suggestion
    return $ helper <$> vf <*> tf
```

Reply:
This is really cool! Didn't know about applicatives before this.